File: crates/claria-core/src/models/findings.rs
//! Durable review findings for one report.
//!
//! A review sweep reads the accepted draft and emits [`Finding`] rows: a style
//! pass proposes an anchored text replacement the user can apply and undo, and
//! a consistency pass reports only. That asymmetry is structural, not a
//! convention — [`ReviewPass::Consistency`] with a [`StyleProposal`] fails
//! validation, so a consistency finding can never grow a write path.
//!
//! Findings are anchored to a [`FindingAnchor`], a `(section_id, revision)`
//! pair. Whether an anchor still describes the section is **derived** by
//! [`finding_is_stale`] rather than written by whoever edited the report; see
//! that function for why.

use std::collections::HashSet;

use jiff::Timestamp;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{
    error::CoreError,
    models::report::{ReportContent, validate_nonempty_text},
};
/// Findings cite records exactly the way a drafting run does, so they share the
/// one definition. Two `RecordCitation` types would also collide in the specta
/// bindings, which refuse duplicate exported names.
pub use crate::models::report_run::{MAX_CITATION_QUOTE_CHARACTERS, RecordCitation};

pub const REPORT_FINDINGS_SCHEMA_VERSION: u32 = 1;

/// Ceiling on findings retained for one report, resolved history included.
pub const MAX_REPORT_FINDINGS: usize = 200;

/// One review-pass coverage receipt per property, so a sweep that produced no
/// findings is still distinguishable from a sweep that never ran.
pub const MAX_REVIEW_COVERAGE: usize = 32;

/// Review property names are host-authored identifiers (`tense_drift`,
/// `unsupported_claim`), not model prose.
pub const MAX_FINDING_PROPERTY_CHARACTERS: usize = 64;

pub const MAX_FINDING_DESCRIPTION_CHARACTERS: usize = 1_000;

/// Record filenames are client-chosen and can be long; this only stops a
/// pathological value from being persisted.
const MAX_RECORD_FILENAME_CHARACTERS: usize = 1_024;

const MAX_MODEL_ID_CHARACTERS: usize = 200;

/// Every review finding recorded against one report, plus the coverage
/// receipts of the sweeps that produced them.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, specta::Type)]
pub struct ReportFindings {
    pub schema_version: u32,
    pub report_id: Uuid,
    pub client_id: Uuid,
    pub findings: Vec<Finding>,
    pub coverage: Vec<ReviewCoverage>,
    #[specta(type = String)]
    pub updated_at: Timestamp,
}

/// One thing a review pass noticed about one section.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, specta::Type)]
pub struct Finding {
    pub id: Uuid,
    pub pass: ReviewPass,
    /// The review property that raised this, e.g. `tense_drift`.
    pub property: String,
    /// The reviewing model, carried so applying a style proposal can credit
    /// the model that wrote the replacement.
    pub model_id: String,
    pub anchor: FindingAnchor,
    pub description: String,
    /// Where in the anchored section the finding points, resolved host-side
    /// from the model's quote.
    pub span: Option<TextSpan>,
    /// The passage this one contradicts. Consistency findings only in
    /// practice; nothing structural forbids a style finding from carrying one.
    pub conflicting: Option<ConflictingRef>,
    pub record_citation: Option<RecordCitation>,
    /// The anchored replacement a style finding offers. Never present on a
    /// consistency finding — see [`ReviewPass::Consistency`].
    pub proposal: Option<StyleProposal>,
    pub status: FindingStatus,
    /// The draft revision the applied replacement produced. Set only while
    /// the finding is [`FindingStatus::Applied`], so an undo can say which
    /// revision it is unwinding.
    pub applied_revision: Option<u64>,
    #[specta(type = Option<String>)]
    pub resolved_at: Option<Timestamp>,
    #[specta(type = String)]
    pub created_at: Timestamp,
}

/// The section and revision a finding describes.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, specta::Type)]
pub struct FindingAnchor {
    pub section_id: Uuid,
    /// The section's authorship revision when the review read it.
    pub revision: u64,
}

/// A character range inside one block of the anchored section. Offsets are
/// `char` indices, not bytes.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, specta::Type)]
pub struct TextSpan {
    /// Zero-based index into the section's blocks.
    pub block_index: u32,
    pub start_char: u32,
    pub end_char: u32,
}

/// The passage a finding conflicts with.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, specta::Type)]
pub struct ConflictingRef {
    /// `None` means the conflicting passage is in the anchored section itself.
    pub section_id: Option<Uuid>,
    pub quote: String,
    pub span: Option<TextSpan>,
}

/// A style pass's anchored replacement: swap `original_text` for
/// `replacement_text` inside one paragraph of the anchored section.
///
/// The match is by text, not offset, and must be unique in the block — a
/// replacement that no longer matches, or matches twice, is refused rather
/// than guessed at.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, specta::Type)]
pub struct StyleProposal {
    /// Zero-based index into the anchored section's blocks.
    pub block_index: u32,
    pub original_text: String,
    pub replacement_text: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, specta::Type)]
#[serde(rename_all = "snake_case")]
pub enum ReviewPass {
    /// Anchored, applicable replacements: tense, terminology, transitions,
    /// redundancy.
    Style,
    /// Read-only: contradictions, unsupported claims, cross-section conflicts.
    /// The pass has no write access, and the validator enforces it.
    Consistency,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, specta::Type)]
#[serde(rename_all = "snake_case")]
pub enum FindingStatus {
    Open,
    /// A style replacement the user applied. Reversible until the surrounding
    /// text changes.
    Applied,
    Dismissed,
    /// The anchored section moved on. Written lazily by the list path; the
    /// authoritative answer is always [`finding_is_stale`].
    Invalidated,
}

/// What the user chose to do with one finding.
///
/// This is the request shape, distinct from [`FindingStatus`], which records
/// where a finding ended up: undoing a finding puts it back to
/// [`FindingStatus::Open`], so the two do not correspond one to one.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, specta::Type)]
#[serde(rename_all = "snake_case")]
pub enum FindingAction {
    ApplyStyle,
    UndoStyle,
    Dismiss,
}

/// Proof that one review property actually ran over one revision, including
/// the case where it found nothing.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, specta::Type)]
pub struct ReviewCoverage {
    pub pass: ReviewPass,
    pub property: String,
    pub model_id: String,
    /// The draft revision the pass read.
    pub revision: u64,
    pub sections_reviewed: u32,
    pub findings: u32,
    #[specta(type = String)]
    pub completed_at: Timestamp,
}

impl ReportFindings {
    /// An empty findings object, which is what a report that has never been
    /// reviewed has.
    pub fn new(client_id: Uuid, report_id: Uuid, now: Timestamp) -> Self {
        Self {
            schema_version: REPORT_FINDINGS_SCHEMA_VERSION,
            report_id,
            client_id,
            findings: Vec::new(),
            coverage: Vec::new(),
            updated_at: now,
        }
    }

    pub fn find(&self, finding_id: Uuid) -> Option<&Finding> {
        self.findings
            .iter()
            .find(|finding| finding.id == finding_id)
    }

    pub fn find_mut(&mut self, finding_id: Uuid) -> Option<&mut Finding> {
        self.findings
            .iter_mut()
            .find(|finding| finding.id == finding_id)
    }

    pub fn validate(&self) -> Result<(), CoreError> {
        if self.schema_version > REPORT_FINDINGS_SCHEMA_VERSION {
            return Err(invalid(format!(
                "review findings schema version {} is newer than supported version {REPORT_FINDINGS_SCHEMA_VERSION}",
                self.schema_version
            )));
        }
        if self.schema_version != REPORT_FINDINGS_SCHEMA_VERSION {
            return Err(invalid(format!(
                "unsupported review findings schema version {}",
                self.schema_version
            )));
        }
        if self.report_id.is_nil() || self.client_id.is_nil() {
            return Err(invalid("review findings IDs must not be nil"));
        }
        if self.findings.len() > MAX_REPORT_FINDINGS {
            return Err(invalid(format!(
                "a report may retain at most {MAX_REPORT_FINDINGS} review findings"
            )));
        }
        let mut ids = HashSet::with_capacity(self.findings.len());
        for finding in &self.findings {
            finding.validate()?;
            if !ids.insert(finding.id) {
                return Err(invalid(format!("duplicate finding ID {}", finding.id)));
            }
            if finding.created_at > self.updated_at {
                return Err(invalid(
                    "a review finding is newer than its findings object",
                ));
            }
        }

        if self.coverage.len() > MAX_REVIEW_COVERAGE {
            return Err(invalid(format!(
                "review coverage may contain at most {MAX_REVIEW_COVERAGE} rows"
            )));
        }
        let mut properties = HashSet::with_capacity(self.coverage.len());
        for coverage in &self.coverage {
            coverage.validate()?;
            if !properties.insert(coverage.property.as_str()) {
                return Err(invalid(format!(
                    "duplicate review coverage for property {}",
                    coverage.property
                )));
            }
        }
        Ok(())
    }
}

impl Finding {
    fn validate(&self) -> Result<(), CoreError> {
        if self.id.is_nil() {
            return Err(invalid("finding ID must not be nil"));
        }
        if self.anchor.section_id.is_nil() {
            return Err(invalid("finding anchor section ID must not be nil"));
        }
        validate_nonempty_text(
            "finding property",
            &self.property,
            MAX_FINDING_PROPERTY_CHARACTERS,
        )?;
        validate_nonempty_text("finding model ID", &self.model_id, MAX_MODEL_ID_CHARACTERS)?;
        validate_nonempty_text(
            "finding description",
            &self.description,
            MAX_FINDING_DESCRIPTION_CHARACTERS,
        )?;
        if let Some(span) = &self.span {
            span.validate()?;
        }
        if let Some(conflicting) = &self.conflicting {
            if conflicting.section_id.is_some_and(|id| id.is_nil()) {
                return Err(invalid("conflicting section ID must not be nil"));
            }
            validate_nonempty_text(
                "finding conflicting quote",
                &conflicting.quote,
                MAX_CITATION_QUOTE_CHARACTERS,
            )?;
            if let Some(span) = &conflicting.span {
                span.validate()?;
            }
        }
        if let Some(citation) = &self.record_citation {
            validate_nonempty_text(
                "finding citation filename",
                &citation.filename,
                MAX_RECORD_FILENAME_CHARACTERS,
            )?;
            validate_nonempty_text(
                "finding citation quote",
                &citation.quote,
                MAX_CITATION_QUOTE_CHARACTERS,
            )?;
        }

        // The consistency pass is read-only by construction. Refusing the
        // combination here means no review sweep, tool schema, or apply path
        // can hand a consistency finding a write.
        if self.pass == ReviewPass::Consistency && self.proposal.is_some() {
            return Err(invalid(
                "a consistency finding must not carry a replacement proposal",
            ));
        }
        if let Some(proposal) = &self.proposal {
            proposal.validate()?;
        }

        match self.status {
            FindingStatus::Applied => {
                if self.pass != ReviewPass::Style {
                    return Err(invalid("only a style finding can be applied"));
                }
                if self.applied_revision.is_none() {
                    return Err(invalid(
                        "an applied finding must record the revision it produced",
                    ));
                }
                if self.proposal.is_none() {
                    return Err(invalid(
                        "an applied finding must retain its replacement proposal",
                    ));
                }
            }
            FindingStatus::Open | FindingStatus::Dismissed | FindingStatus::Invalidated => {
                if self.applied_revision.is_some() {
                    return Err(invalid(
                        "only an applied finding may record an applied revision",
                    ));
                }
            }
        }
        // An unresolved finding has no resolution timestamp, and every
        // resolved one has exactly one. Undo therefore has to clear both
        // fields, not just the revision.
        match (self.status, self.resolved_at) {
            (FindingStatus::Open, Some(_)) => {
                Err(invalid("an open finding must not record a resolution time"))
            }
            (FindingStatus::Open, None) => Ok(()),
            (_, None) => Err(invalid("a resolved finding must record when")),
            (_, Some(resolved_at)) if resolved_at < self.created_at => {
                Err(invalid("a finding was resolved before it was created"))
            }
            (_, Some(_)) => Ok(()),
        }
    }
}

impl TextSpan {
    fn validate(&self) -> Result<(), CoreError> {
        if self.end_char <= self.start_char {
            return Err(invalid("a text span must end after it starts"));
        }
        Ok(())
    }
}

impl StyleProposal {
    fn validate(&self) -> Result<(), CoreError> {
        validate_nonempty_text(
            "style proposal original text",
            &self.original_text,
            MAX_CITATION_QUOTE_CHARACTERS,
        )?;
        validate_nonempty_text(
            "style proposal replacement text",
            &self.replacement_text,
            MAX_CITATION_QUOTE_CHARACTERS,
        )?;
        if self.original_text == self.replacement_text {
            return Err(invalid("a style proposal must change the text"));
        }
        Ok(())
    }
}

impl ReviewCoverage {
    fn validate(&self) -> Result<(), CoreError> {
        validate_nonempty_text(
            "review coverage property",
            &self.property,
            MAX_FINDING_PROPERTY_CHARACTERS,
        )?;
        validate_nonempty_text(
            "review coverage model ID",
            &self.model_id,
            MAX_MODEL_ID_CHARACTERS,
        )?;
        Ok(())
    }
}

/// Decode and validate persisted review findings. Validation is part of
/// decoding, exactly as it is for the workspace object, so a corrupt or
/// future-versioned object is never presented as review state.
pub fn decode_report_findings(bytes: &[u8]) -> Result<ReportFindings, CoreError> {
    let findings: ReportFindings = serde_json::from_slice(bytes)?;
    findings.validate()?;
    Ok(findings)
}

/// Whether `finding`'s anchor still describes the report.
///
/// Staleness is derived on every read and never written by a mutator. Nothing
/// that edits a report — a hand save, a proposal accept, a drafting run, an
/// applied replacement, a revert — has to remember to walk the findings and
/// invalidate them, so no mutation path can forget to. The
/// [`FindingStatus::Invalidated`] stamp the list path writes is a cache of
/// this answer for display; it is never the source of truth.
///
/// A finding is stale when its section is gone, or when the section's
/// authorship stamp has moved past the revision the review read. A section
/// with no authorship stamp has not been rewritten since the review, so it is
/// fresh.
pub fn finding_is_stale(finding: &Finding, content: &ReportContent) -> bool {
    match content
        .sections
        .iter()
        .find(|section| section.id == finding.anchor.section_id)
    {
        None => true,
        Some(section) => section
            .authorship
            .as_ref()
            .is_some_and(|authorship| authorship.revision > finding.anchor.revision),
    }
}

fn invalid(message: impl Into<String>) -> CoreError {
    CoreError::InvalidFindings(message.into())
}

File: claria-desktop-frontend/src/lib/findings.ts
import type {
  CompletionCheckKind,
  CompletionReport,
  Finding,
  ReportContent,
} from "./tauri";

/** Which findings the list is showing. */
export type FindingsFilter = "all" | "style" | "consistency" | "resolved";

/**
 * How one finding is drawn. Derived on every render rather than read off
 * `finding.status` alone: the backend treats staleness as a question it
 * answers from the report, and the stored `invalidated` stamp is only a cache
 * of that answer. A finding whose section moved on since the review is
 * invalid here even if nothing has written the stamp yet.
 */
export type FindingState = "open" | "applied" | "dismissed" | "invalidated";

export type FindingGroup = {
  sectionId: string;
  /** The section's current heading, or the removed-section stand-in. */
  heading: string;
  /** Findings still open in this section, counted before any filtering. */
  openCount: number;
  findings: Finding[];
};

/** The section a finding points at is no longer in the report. */
const REMOVED_SECTION_HEADING = "Section no longer in the report";

/**
 * The backend names review properties in snake_case, and that is the name the
 * findings themselves carry. Rendering it as prose is presentation, so it
 * happens here rather than being sent over the bridge twice.
 */
export function reviewPropertyLabel(property: string): string {
  return property.replaceAll("_", " ");
}

/**
 * Whether the finding's anchor still describes the report — the frontend's
 * copy of the backend's `finding_is_stale`. A section that is gone is stale,
 * and so is one whose authorship stamp has moved past the revision the review
 * read. A section with no stamp has not been rewritten, so it is fresh.
 */
export function findingIsStale(
  finding: Finding,
  content: ReportContent
): boolean {
  const section = content.sections.find(
    (candidate) => candidate.id === finding.anchor.section_id
  );
  if (!section) return true;
  const authorship = section.authorship;
  return authorship ? authorship.revision > finding.anchor.revision : false;
}

/**
 * Resolution wins over staleness: an applied replacement keeps its receipt so
 * the undo stays reachable, and a dismissed finding stays dismissed. Only an
 * open finding can go stale.
 */
export function findingState(
  finding: Finding,
  content: ReportContent
): FindingState {
  if (finding.status === "applied") return "applied";
  if (finding.status === "dismissed") return "dismissed";
  if (finding.status === "invalidated") return "invalidated";
  return findingIsStale(finding, content) ? "invalidated" : "open";
}

/** The revision the anchored section is on now, for the staleness note. */
export function currentSectionRevision(
  finding: Finding,
  content: ReportContent,
  draftRevision: number
): number {
  const section = content.sections.find(
    (candidate) => candidate.id === finding.anchor.section_id
  );
  return section?.authorship?.revision ?? draftRevision;
}

export function matchesFilter(
  finding: Finding,
  state: FindingState,
  filter: FindingsFilter
): boolean {
  switch (filter) {
    case "all":
      return true;
    case "style":
      return finding.pass === "style";
    case "consistency":
      return finding.pass === "consistency";
    case "resolved":
      return state === "applied" || state === "dismissed";
  }
}

/**
 * Findings grouped by section in document order, filtered for display.
 *
 * `openCount` is counted before the filter runs, so the "Style" chip does not
 * make a section with three open consistency flags look clean. Findings whose
 * section has been deleted collect in one trailing group rather than
 * disappearing — a finding nobody can see is a finding nobody can dismiss.
 */
export function groupFindings(
  findings: readonly Finding[],
  content: ReportContent,
  filter: FindingsFilter
): FindingGroup[] {
  const openCounts = openFindingCounts(findings, content);
  const groups = new Map<string, FindingGroup>();
  for (const section of content.sections) {
    groups.set(section.id, {
      sectionId: section.id,
      heading: section.heading,
      openCount: openCounts.get(section.id) ?? 0,
      findings: [],
    });
  }
  const removed: FindingGroup = {
    sectionId: "",
    heading: REMOVED_SECTION_HEADING,
    openCount: 0,
    findings: [],
  };
  for (const finding of findings) {
    if (!matchesFilter(finding, findingState(finding, content), filter)) {
      continue;
    }
    (groups.get(finding.anchor.section_id) ?? removed).findings.push(finding);
  }
  const ordered = [...groups.values()].filter(
    (group) => group.findings.length > 0
  );
  if (removed.findings.length > 0) ordered.push(removed);
  return ordered;
}

/**
 * Open findings per section, for the canvas flag chips. Derived from the
 * findings themselves rather than from run state: a review can be long over
 * by the time the reader looks at the document.
 */
export function openFindingCounts(
  findings: readonly Finding[],
  content: ReportContent
): ReadonlyMap<string, number> {
  const counts = new Map<string, number>();
  for (const finding of findings) {
    if (findingState(finding, content) !== "open") continue;
    const sectionId = finding.anchor.section_id;
    counts.set(sectionId, (counts.get(sectionId) ?? 0) + 1);
  }
  return counts;
}

/**
 * The passage the finding points at, sliced out of the anchored section.
 *
 * `null` when the span is missing or no longer addresses a paragraph — the
 * card then shows the description alone rather than an invented quote.
 */
export function anchoredQuote(
  finding: Finding,
  content: ReportContent
): string | null {
  const span = finding.span;
  if (!span) return null;
  const section = content.sections.find(
    (candidate) => candidate.id === finding.anchor.section_id
  );
  const block = section?.blocks[span.block_index];
  if (!block || block.kind !== "paragraph") return null;
  const quote = block.text.slice(span.start_char, span.end_char).trim();
  return quote === "" ? null : quote;
}

/** The heading of the section a consistency finding conflicts with. */
export function conflictingHeading(
  finding: Finding,
  content: ReportContent
): string | null {
  const sectionId = finding.conflicting?.section_id;
  if (!sectionId) return null;
  return (
    content.sections.find((candidate) => candidate.id === sectionId)?.heading ??
    null
  );
}

export type CompletionSummaryRow = {
  kind: CompletionCheckKind;
  /** One line, already counted and pluralised. */
  label: string;
  /** Headings of the sections at fault, in document order. */
  sections: string[];
};

const COMPLETION_LABELS: Record<
  CompletionCheckKind,
  (count: number) => string
> = {
  section_not_terminal: (count) =>
    count === 1
      ? "1 section never finished drafting"
      : `${count} sections never finished drafting`,
  required_section_empty: (count) =>
    count === 1
      ? "1 required section is still empty"
      : `${count} required sections are still empty`,
  unresolved_citation: (count) =>
    count === 1
      ? "1 citation could not be verified"
      : `${count} citations could not be verified`,
  missing_citation: (count) =>
    count === 1
      ? "1 section cites no records"
      : `${count} sections cite no records`,
  placeholder_text: (count) =>
    count === 1
      ? "1 section still has placeholder text"
      : `${count} sections still have placeholder text`,
  unresolved_finding: (count) =>
    count === 1 ? "1 open finding" : `${count} open findings`,
};

/** The order the checklist reads in, worst-structured problem first. */
const COMPLETION_ORDER: CompletionCheckKind[] = [
  "section_not_terminal",
  "required_section_empty",
  "placeholder_text",
  "missing_citation",
  "unresolved_citation",
  "unresolved_finding",
];

/**
 * One line per failing check kind, counted and named. Checks with no section
 * (a placeholder left in the title, say) contribute to the count without
 * naming anything.
 */
export function summarizeCompletion(
  report: CompletionReport,
  content: ReportContent
): CompletionSummaryRow[] {
  const headings = new Map(
    content.sections.map((section) => [section.id, section.heading])
  );
  const byKind = new Map<CompletionCheckKind, string[]>();
  const counts = new Map<CompletionCheckKind, number>();
  for (const check of report.checks) {
    counts.set(check.kind, (counts.get(check.kind) ?? 0) + 1);
    const heading = check.section_id
      ? headings.get(check.section_id)
      : undefined;
    if (heading === undefined) continue;
    const named = byKind.get(check.kind) ?? [];
    if (!named.includes(heading)) named.push(heading);
    byKind.set(check.kind, named);
  }
  return COMPLETION_ORDER.filter((kind) => counts.has(kind)).map((kind) => ({
    kind,
    label: COMPLETION_LABELS[kind](counts.get(kind) ?? 0),
    sections: byKind.get(kind) ?? [],
  }));
}
